homeassistant/helpers/entity.py

```diff
@@ -736,3 +746,53 @@ async def async_toggle(self, **kwargs: Any) -> None:
             await self.async_turn_off(**kwargs)
         else:
             await self.async_turn_on(**kwargs)
+
+
+class MeasurableUnitEntity(Entity):
```

Comment:
Is the plan to create a measurable entity that can handle any unit conversion or should we consider mixins per unit conversion (ie `TemperatureEntity`)

Reply:
I think the former, it's fairly common pattern for integrations to have a shared class for all sensors, then just set the attributes, including `unit_of_measurement`, based on some input from the lib.
If we want to specialize like this, those integrations would have to be refactored.

Reply:
I agree. Besides temperature, I think we should only convert units based on device class. That will allow us to have a look up table for the conversion to apply and a clean implementation in a single class.

Reply:
> we should only convert units based on device class

I actually implemented it like that first, but there are two issues:

1. We will need to be backwards compatible for some time to handle integrations which don't set the device class.
2. How to handle integrations such as `input_number` which allow configuring a `unit_of_measurement` but not a `device_class`?

The first issue is not a problem, but I'm not sure about the second one.

Reply:
If the unit of measurement is configured by the user we probably shouldn't convert it.

Reply:
OK, that makes sense for `input_number`.
Not doing the conversion could be considered a breaking change since we currently do convert all temperatures, but maybe it would be for the better?

`unit_of_measurement` is also used by base components `geo_location` (for distance) and `air_quality` (concentrations, I think).

Reply:
I really don't think device class and unit of measurement should be coupled. I think in the long run it would make more sense to device a measurement class/type.

We are more and more mixing up the object and the measurement it provides. For example, temperature vs garage door.

In the long run, this is going to bite.

Reply:
device class and unit of measurement need to be coupled for blueprints and to export it to Homekit/Google/Alexa. It's not possible to make code that can work with everything. If it's a device class, we can categorize the entity and know how to deal with it.